peppy/adapters/ip2location.py

```diff
@@ -0,0 +1,67 @@
+from __future__ import annotations
+
+from dataclasses import dataclass
+
+import requests
+
+
+# Would much prefer Pydantic.
+@dataclass
+class IPQueryResult:
+    ip: str
+    country_code: str
+    country_name: str
+    region_name: str
+    latitude: float
+    longitude: float
+    is_proxy: bool
+
+
+IP2LOCATION_BASE_API = "https://api.ip2location.io/"
+"""The default root endpoint for the IP2Location API."""
+
+
+class GeolocationException(Exception):
+    ...
+
+
+class Ip2LocationApi:
+    def __init__(
+        self,
+        api_key: str,
+        *,
+        api_root_url: str = IP2LOCATION_BASE_API,
+        silent_fail: bool = False,
+    ) -> None:
+        self.root_url = api_root_url
+        self.api_key = api_key
+        self.silent_fail = silent_fail
+
+    def query_ip(self, ip_address: str) -> IPQueryResult | None:
+        query_response = requests.get(
+            self.root_url,
+            params={
+                "key": self.api_key,
+                "ip": ip_address,
+            },
+        ).json()
+
+        if "error" in query_response:
```

Comment:
you might want to also check the status code just in case they have an error in the proxy or something - i don't think `requests` will automatically raise if the status is incorrect right? you also might not wanna jump straight to `.json` for the same reason. i think there's a `raise_for_status_code` or similar method on the response object?